Add section about extensible variant types. (#211)
rescript-association/rescript-lang.org: pages/docs/manual/latest/extensible-variant.mdx

---
title: "Extensible Variant"
description: "Extensible Variants in ReScript"
canonical: "/docs/manual/latest/extensible-variant"
---

> # Extensible Variant Types

Reviewer: ```suggestion
# Extensible Variant
```

> Variants can be made extensible by defining a type using `..` and adding new variant constructors using `+=`.

Reviewer: ```suggestion
Variant types are usually constraint to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type (e.g. in different places of your program). For this, we offer Extensible Variant types.

## Definition and Usage
```

<CodeTab labels={["ReScript", "JS Output"]}>

```res example
type t = ..

type t += Other

type t +=
  | Point(float, float)
  | Line(float, float, float, float)
```
```js
var Caml_exceptions = require("./stdlib/caml_exceptions.js");

var Other = Caml_exceptions.create("Playground.Other");

var Point = Caml_exceptions.create("Playground.Point");

var Line = Caml_exceptions.create("Playground.Line");
```

</CodeTab>

> 

Reviewer: ```suggestion

The `..` in the type declaration above defines an extensible variant type. The `+=` operator is used to add constructors to an Extensible Variant type. 

```

> Pattern-matching is possible the same way as with normal variants but there is one caveat:

Reviewer: ```suggestion
## Pattern Matching Caveats

Extensible Variants are open-ended, so the compiler will not be able to exhaustively pattern match all available cases. You will always need to provide a default `_` case for every `switch` expression. 
```

> With extensible variants, the possibility to check for exhaustiveness vanishes, you always need a default case `_` here.

Reviewer: ```suggestion
```

<CodeTab labels={["ReScript", "JS Output"]}>

```res example
let print = v =>
  switch v {
  | Point(x, y) => Js.log2("Point", (x, y))
  | Line(ax, ay, bx, by) => Js.log2("Line", (ax, ay, bx, by))
  | Other
  | _ => Js.log("Other")
  }
```
```js
function print(v) {
  if (v.RE_EXN_ID === Point) {
    console.log("Point", [v._1, v._2]);
  } else if (v.RE_EXN_ID === Line) {
    console.log("Line", [v._1, v._2, v._3, v._4]);
  } else {
    console.log("Other");
  } 
}
```

> </CodeTab>

Reviewer: ```suggestion
</CodeTab>

## Tips & Tricks

**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as Extensible Variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`.

It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

---
title: "Extensible Variant"
description: "Extensible Variants in ReScript"
canonical: "/docs/manual/latest/extensible-variant"
---

# Extensible Variant

> Variant types are usually constrained to a fixed set of constructors. There may be very rare cases where you still want to be able to add constructors to a variant type (even after its initial type declaration). For this, we offer extensible variant types.

Reviewer: ```suggestion
Variant types are usually constrained to a fixed set of constructors. There may be very rare cases where you still want to be able to add constructors to a variant type even after its initial type declaration. For this, we offer extensible variant types.
```

> Variant types are usually constraint to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type (e.g. in different places of your program). For this, we offer extensible variant types.

Author: are usually **constrained** ...

> Variant types are usually constraint to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type (e.g. in different places of your program). For this, we offer extensible variant types.

Reviewer: ```suggestion
Variant types are usually constrained to a fixed set of constructors that are defined in its type declaration. There may be very rare cases where you still want to be able to add constructors to a variant type, even after the initial type declaration (e.g. in different places of your program). For this, we offer extensible variant types.
```

Author: Perfect!

## Definition and Usage

<CodeTab labels={["ReScript", "JS Output"]}>

```res example
type t = ..

type t += Other

type t +=
  | Point(float, float)
  | Line(float, float, float, float)
```
```js
var Caml_exceptions = require("./stdlib/caml_exceptions.js");

var Other = Caml_exceptions.create("Playground.Other");

var Point = Caml_exceptions.create("Playground.Point");

var Line = Caml_exceptions.create("Playground.Line");
```

</CodeTab>

The `..` in the type declaration above defines an extensible variant `type t`. The `+=` operator is then used to add constructors to the given type. 

## Pattern Matching Caveats

Extensible variants are open-ended, so the compiler will not be able to exhaustively pattern match all available cases. You will always need to provide a default `_` case for every `switch` expression. 

<CodeTab labels={["ReScript", "JS Output"]}>

> ```res example

Reviewer: ```suggestion
```res
```

let print = v =>
  switch v {
  | Point(x, y) => Js.log2("Point", (x, y))
  | Line(ax, ay, bx, by) => Js.log2("Line", (ax, ay, bx, by))
  | Other
  | _ => Js.log("Other")
  }
```
```js
function print(v) {
  if (v.RE_EXN_ID === Point) {
    console.log("Point", [v._1, v._2]);
  } else if (v.RE_EXN_ID === Line) {
    console.log("Line", [v._1, v._2, v._3, v._4]);
  } else {
    console.log("Other");
  } 
}
```

</CodeTab>

## Tips & Tricks

> **Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were extensible variants make sense. We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.

Reviewer: ```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's one of the very few use-case where extensible variants make sense.

We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

> **Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as Extensible Variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.

Reviewer: ```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

> **Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were Extensible Variants make sense. We usually recommend sticking with common [Variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.

Reviewer: ```suggestion
**Fun fact:** In ReScript, [exceptions](./exception) are actually implemented as extensible variants under the hood, so `exception UserError(string)` is equivalent to `type exn += UserError(string)`. It's probably one of the few use-cases were extensible variants make sense. We usually recommend sticking with common [variants](./variant) as much as possible to reap the benefits of exhaustive pattern matching.
```

---
title: "Extensible Variant"
description: "Extensible Variants in ReScript"
canonical: "/docs/manual/latest/extensible-variant"
---

# Extensible Variant

Variant types are usually constrained to a fixed set of constructors. There may be very rare cases where you still want to be able to add constructors to a variant type even after its initial type declaration. For this, we offer extensible variant types.

## Definition and Usage

<CodeTab labels={["ReScript", "JS Output"]}>

```res example
type t = ..

type t += Other

type t +=
  | Point(float, float)
  | Line(float, float, float, float)
```
```js
var Caml_exceptions = require("./stdlib/caml_exceptions.js");

var Other = Caml_exceptions.create("Playground.Other");

var Point = Caml_exceptions.create("Playground.Point");

var Line = Caml_exceptions.create("Playground.Line");
```

</CodeTab>

> The `..` in the type declaration above defines an extensible variant `type t`. The `+=` operator is then used to add constructors to the given type. 

Reviewer: ```suggestion
The `..` in the type declaration above defines an extensible variant `type t`. The `+=` operator is then used to add constructors to the given type. 

**Note:** Don't forget the leading `type` keyword when using the `+=` operator!
```
